fsharp-backend/src/LibExecution/Interpreter.fs

```diff
@@ -278,16 +282,39 @@ let rec eval' (state : ExecutionState) (st : Symtable) (e : Expr) : DvalTask =
             let (passes, newVars, traces) = checkPattern v p
 
             if passes then
-              true, newVars, [ (id, dv) ] @ traces
+              true, newVars, (id, dv) :: traces
             else
-              false,
-              newVars,
-              (id, incomplete id) :: traceArgsIncomplete [ p ] @ traces
+              false, newVars, traceIncompleteWithArgs id [ p ] @ traces
 
           | _name, argPatterns, _dv ->
-            let traces = (id, incomplete id) :: traceArgsIncomplete argPatterns
+            let traces = traceIncompleteWithArgs id argPatterns
             false, [], traces
 
+        | PTuple (id, firstPat, secondPat, theRestPat) ->
+          let allPatterns = firstPat :: secondPat :: theRestPat
+
+          match dv with
+          | DTuple (first, second, theRest) ->
+            let allVals = first :: second :: theRest
+
+            if List.length allVals = List.length allPatterns then
+              let (passResults, newVarResults, traceResults) =
+                List.zip allVals allPatterns
+                |> List.map (fun (dv, pat) -> checkPattern dv pat)
+                |> List.unzip3
+
+              let allPass = passResults |> List.forall (fun arg -> arg)
```

Comment:
There should be a function called `identity` somewhere that we can use for this.

Reply:
Found it - `Fun.identity`!